### PR TITLE
SPARKX on diet

### PR DESCRIPTION
**Pull request by NGoetz**
Completely new approach of saving data. Instead of having dozens of float/int objects, everything is saved in a numpy array. This saves around 30% of RAM and some runtime on setup. 
This has two side effects: Before we used `None` objects for unset values. Now we use `nan`. All `ints` are now `floats`. The user will have to deal with this, and cast accordingly.
This is a very critical change, so I would ask both of you to have a look and catch any bug or unexpected side effect.
Please also check if you like what I did to the documentation.

### CODE REVIEW

File: src/sparkx/Particle.py [hunk 1]
```diff
@@ -8,149 +8,159 @@ class Particle:
...
+    strangeness : short
         Strangeness quantum number of the particle.
-    weight_ : float
+    weight : float
         Weight of the particle.
 
+    These attributes are saved in following data structure:
+
+    Data
+    -----
```
Comment (Hendrik1704) [inline src/sparkx/Particle.py:72]: One "-" too much.

File: src/sparkx/Particle.py [hunk 2]
```diff
@@ -267,130 +243,131 @@ def __initialize_from_array(self,input_format,particle_array):
...
-                "form_time_": 13,
-                "xsecfac_": 14,
-                "proc_id_origin_": 15,
-                "proc_type_origin_": 16,
-                "t_last_coll_": 17,
-                "pdg_mother1_": 18,
-                "pdg_mother2_": 19,
-                "baryon_number_": 20,
-                "strangeness_": 21,
+                 "t_": [0,0],
```
Comment (Hendrik1704) [inline src/sparkx/Particle.py:288]: One whitespace too much in the beginning.

File: src/sparkx/Particle.py [hunk 3]
```diff
@@ -267,130 +243,131 @@ def __initialize_from_array(self,input_format,particle_array):
...
-                        setattr(self,attribute,None)
+                    if len(particle_array)<index[1]+1:
                         continue
                     # Type casting for specific attributes
                     if attribute in ["t_", "x_", "y_", "z_", "mass_", "E_", "px_", "py_", "pz_", "form_time_", "xsecfac_", "t_last_coll_", "weight_"]:
-                        setattr(self, attribute, float(particle_array[index]))
-                    elif attribute in ["pdg_", "ID_", "charge_", "ncoll_", "proc_id_origin_", "proc_type_origin_", "pdg_mother1_", "pdg_mother2_", "baryon_number_", "strangeness_", "status_"]:
-                        setattr(self, attribute, int(particle_array[index]))
+                        self.data_[index[0]] = float(particle_array[index[1]])
+                    elif attribute in ["pdg_", "ID_", "ncoll_", "proc_id_origin_", "proc_type_origin_", "pdg_mother1_", "pdg_mother2_", "status_"]:
```
Comment (Hendrik1704) [inline src/sparkx/Particle.py:354]: Why is this still present? I thought that everything is now stored as `float`. Then we don't need the casts.

File: src/sparkx/Particle.py [hunk 4]
```diff
@@ -688,55 +664,69 @@ def status(self):
...
+        return self.data_[24]
 
     @weight.setter
     def weight(self,value):
-        self.weight_ = value
+        self.data_[24] = value
+
+    @property
+    def pdg_valid(self):
+        """Get the weight of the particle.
```
Comment (Hendrik1704) [inline src/sparkx/Particle.py:719]: This is for sure not the purpose of this function ;-)

File: src/sparkx/Oscar.py [hunk 5]
```diff
@@ -594,7 +594,7 @@ def uncharged_particles(self):
 
         for i in range(0, self.num_events_):
             self.particle_list_[i] = [elem for elem in self.particle_list_[i]
-                                        if (elem.charge == 0 and elem.charge != None)]
+                                        if (elem.charge == 0 and elem.charge != np.nan)]
```
Comment (Hendrik1704) [inline src/sparkx/Oscar.py:597]: I just saw that there is also a `and elem.is_strange() != np.nan` needed in line 616 in the `strange_particles` filter.

**Review by Hendrik1704 — CHANGES_REQUESTED** [review]
I went through the code and added some small comments. There is also the point with the cast in the setting of particles from arrays function which I don't understand completely.
I also found a missing `np.nan` check in the Oscar and Jetscape class. See comments.

I also tried to follow the indices in the array. Didn't spot any mistakes. I guess we should still run the test from PR #158 to see if we get any errors there.

File: src/sparkx/Particle.py (hunk 3 above)
Comment (NGoetz) [reply src/sparkx/Particle.py:354]: That is a good point indeed. I added it to ensure that only only ints, even though in the format of a float, are stored. Imagine the user is a madman, feeding a file with 2.5 charge - how should we act? Allow the 2.5 charge, or emulate old behaviour by casting it to 2.0?

Comment (Hendrik1704) [reply src/sparkx/Particle.py:354]: Ok, then maybe add a comment there for other people who might stumble across this.

File: src/sparkx/Jetscape.py [hunk 6]
```diff
@@ -645,7 +645,7 @@ def lower_event_energy_cut(self,minimum_event_energy):
```
[the remark this answers is not in the document]
Comment (NGoetz) [reply src/sparkx/Jetscape.py:645]: Ah you meant 644. Got it.

File: src/sparkx/Particle.py (hunk 3 above)
Comment (NGoetz) [reply src/sparkx/Particle.py:354]: Good idea, I will add the comment.

**Comment by NGoetz** [comment]
I have included all the requested changes, but I would first rebase on the sparkx_devel after the test PR merged in order to test this class properly.

**Comment by Hendrik1704** [comment]
The branch with the test is merged to sparkx_devel. You can go ahead with the tests.

**Comment by NGoetz** [comment]
I fixed issues both in the class, the tests and the Changelog.

File: tests/test_Particle.py [hunk 7]
```diff
@@ -54,8 +54,9 @@ def test_pdg():
     # check valid particle
     p = Particle()
     p.pdg = 211
+    print(p.pdg)
```
Comment (Hendrik1704) [inline tests/test_Particle.py:57]: The print should probably not be there.

**Review by Hendrik1704 — APPROVED** [review]
Looks good now

**Pull request merged**